File: 04_BinarySearch/TimeBasedKeyValueStore/solution.py
'''
timestampは単調増加するため、
setの際にタイムスタンプと値のペアをリストの末尾に追加すれば、
getの際には二分探索を行うことで、値をO(logN)で取得できる。
'''

class TimeMap:

    def __init__(self):
        self.map = {} # key: int, value: List[str]  
                
    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.map:
            self.map[key] = []
            self.map[key].append([timestamp, value])
        else:
            self.map[key].append([timestamp, value])
            
    def get(self, key: str, timestamp: int) -> str:
        if key not in self.map:
            return ""
        
        values = self.map[key]
        left, right = 0, len(values) - 1
        while left <= right:
            mid = left + (right - left) // 2
            if values[mid][0] <= timestamp:
                left = mid + 1
            else:
                right = mid - 1       
        return "" if right < 0 else values[right][1]

# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: 04_BinarySearch/TimeBasedKeyValueStore/solution.go
type TimeValuePair struct {
	timestamp int
	value     string
}

type TimeMap struct {
	store map[string][]TimeValuePair
}

func Constructor() TimeMap {
	return TimeMap{store: make(map[string][]TimeValuePair)}
}

func (this *TimeMap) Set(key string, value string, timestamp int) {
	if _, ok := this.store[key]; !ok {
		this.store[key] = make([]TimeValuePair, 0)
	}
	this.store[key] = append(this.store[key], TimeValuePair{timestamp, value})
}

func (this *TimeMap) Get(key string, timestamp int) string {
	// keyが存在しない場合
	if _, ok := this.store[key]; !ok {
		return ""
	}

	// keyが存在する場合は二分探索を行う
	left, right := 0, len(this.store[key])-1
	for left <= right {
		mid := left + (right-left)/2
		if this.store[key][mid].timestamp <= timestamp {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	// 指定されたtimestampが、そのkeyに設定されたすべてのtimestampよりも小さい場合
	if right < 0 {
		return ""
	}

	return this.store[key][right].value
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Set(key,value,timestamp);
 * param_2 := obj.Get(key,timestamp);
 */